<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [55]:
import pandas as pd
import psycopg2

In [54]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [4]:
query_3_1 = f'''select count(id)
            from public.vacancies
            '''

In [5]:
df_3_1 = pd.read_sql_query(query_3_1, connection)
df_3_1

C:\Users\kab27\AppData\Local\Temp\ipykernel_12832\2127050470.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_3_1 = pd.read_sql_query(query_3_1, connection)


count
0  49197

2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [6]:
query_3_2 = f'''select count(id)
            from public.employers
            '''

In [7]:
df_3_2 = pd.read_sql_query(query_3_2, connection)
df_3_2

C:\Users\kab27\AppData\Local\Temp\ipykernel_12832\2741247974.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_3_2 = pd.read_sql_query(query_3_2, connection)


count
0  23501

3. Посчитате с помощью запроса количество регионов (таблица areas).

In [8]:
query_3_3 = f'''select count(id)
            from public.areas
            '''

In [9]:
df_3_3 = pd.read_sql_query(query_3_3, connection)
df_3_3

C:\Users\kab27\AppData\Local\Temp\ipykernel_12832\3006110907.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_3_3 = pd.read_sql_query(query_3_3, connection)


count
0   1362

4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [10]:
query_3_4 = f'''select count(id)
            from public.industries
            '''

In [11]:
df_3_4 = pd.read_sql_query(query_3_4, connection)
df_3_4

C:\Users\kab27\AppData\Local\Temp\ipykernel_12832\764330752.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_3_4 = pd.read_sql_query(query_3_4, connection)


count
0    294

***

Количество вакансий более чем в 2 раза больше чем количество работадателей. Следовательно в основном работадатели публикуют больше 1 вакансии, возможно есть работадатели без вакансий.
Работадатели размещают свои вакансии в 1362х регионах. Всего указано 294 сферы деятельности.

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [12]:
query_4_1 = f'''select 
                a.name,
                count(v.id)
            from public.areas a
                LEFT JOIN  public.vacancies v on v.area_id=a.id
            GROUP BY a.name
            ORDER BY 2 desc
            LIMIT 5
            '''

In [13]:
df_4_1 = pd.read_sql_query(query_4_1, connection)
df_4_1

C:\Users\kab27\AppData\Local\Temp\ipykernel_12832\1522353868.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_4_1 = pd.read_sql_query(query_4_1, connection)


name  count
0           Москва   5333
1  Санкт-Петербург   2851
2            Минск   2112
3      Новосибирск   2006
4           Алматы   1892

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [14]:
query_4_2 = f'''select 
                count(id)
            from public.vacancies
            WHERE salary_from is not null or salary_to is not null
            '''

In [15]:
df_4_2 = pd.read_sql_query(query_4_2, connection)
df_4_2

C:\Users\kab27\AppData\Local\Temp\ipykernel_12832\3761913292.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_4_2 = pd.read_sql_query(query_4_2, connection)


count
0  24073

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [16]:
query_4_3 = f'''select 
                ROUND(AVG(salary_from)) as "нижняя граница",
                ROUND(AVG(salary_to)) as "верхняя граница"
            from public.vacancies
            '''

In [17]:
df_4_3 = pd.read_sql_query(query_4_3, connection)
df_4_3

C:\Users\kab27\AppData\Local\Temp\ipykernel_12832\4169514055.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_4_3 = pd.read_sql_query(query_4_3, connection)


нижняя граница  верхняя граница
0         71065.0         110537.0

4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [18]:
query_4_4 = f'''select
                schedule,
                employment,
                count(id)
            from public.vacancies
            GROUP BY schedule, employment
            ORDER BY 3 desc
            '''

In [19]:
df_4_4 = pd.read_sql_query(query_4_4, connection)
df_4_4

C:\Users\kab27\AppData\Local\Temp\ipykernel_12832\1636312543.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_4_4 = pd.read_sql_query(query_4_4, connection)


schedule           employment  count
0        Полный день     Полная занятость  35367
1   Удаленная работа     Полная занятость   7802
2      Гибкий график     Полная занятость   1593
3   Удаленная работа  Частичная занятость   1312
4     Сменный график     Полная занятость    940
5        Полный день           Стажировка    569
6     Вахтовый метод     Полная занятость    367
7        Полный день  Частичная занятость    347
8      Гибкий график  Частичная занятость    312
9        Полный день     Проектная работа    141
10  Удаленная работа     Проектная работа    133
11     Гибкий график           Стажировка    116
12    Сменный график  Частичная занятость    101
13  Удаленная работа           Стажировка     64
14     Гибкий график     Проектная работа     18
15    Сменный график           Стажировка     12
16    Вахтовый метод     Проектная работа      2
17    Сменный график     Проектная работа      1

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [20]:
query_4_5 = f'''select
                experience,
                count(id)
            from public.vacancies
            GROUP BY experience
            ORDER BY 2
            '''

In [21]:
df_4_5 = pd.read_sql_query(query_4_5, connection)
df_4_5

C:\Users\kab27\AppData\Local\Temp\ipykernel_12832\802873672.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_4_5 = pd.read_sql_query(query_4_5, connection)


experience  count
0         Более 6 лет   1337
1           Нет опыта   7197
2       От 3 до 6 лет  14511
3  От 1 года до 3 лет  26152

***

Чем крупнее город тем больше вакансий расположено в этом регионе.
В половине вакансий не указан уровень зарплаты.
Больше всего вакансий с графиком работы - "Полный день" и тим трудоустройства "Полная занятость".
Наиболее востребованы специалисты с опытом работы от 1 года до 3 лет.

# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [23]:
query_5_1 = f'''(select
                e.name,
                count(v.id) vacancies
            from public.employers e
                JOIN public.vacancies v on v.employer_id = e.id
            GROUP BY e.name
            ORDER BY count(v.id) desc
            limit 1)
            UNION all
            (select
                e.name,
                count(v.id) vacancies
            from public.employers e
                JOIN public.vacancies v on v.employer_id = e.id
            GROUP BY e.name
            ORDER BY count(v.id) desc
            OFFSET 4 limit 1)
            '''

In [24]:
df_5_1 = pd.read_sql_query(query_5_1, connection)
display(df_5_1)

C:\Users\kab27\AppData\Local\Temp\ipykernel_12832\2705250063.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_5_1 = pd.read_sql_query(query_5_1, connection)


name  vacancies
0         Яндекс       1933
1  Газпром нефть        331

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [25]:
query_5_2 = f'''select
                a.name,
                count(e.id) employesrs,
                count(v.id) vacancies
            from public.areas a
                LEFT JOIN public.employers e on e.area = a.id
                LEFT JOIN public.vacancies v on v.area_id = a.id
            WHERE v.id is null
            GROUP BY a.name
            ORDER BY count(e.id) desc
            '''

In [26]:
df_5_2 = pd.read_sql_query(query_5_2, connection)
display(df_5_2)

C:\Users\kab27\AppData\Local\Temp\ipykernel_12832\2401764445.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_5_2 = pd.read_sql_query(query_5_2, connection)


name  employesrs  vacancies
0                            Россия         410          0
1                         Казахстан         207          0
2                Московская область          75          0
3                Краснодарский край          19          0
4                          Беларусь          18          0
..                              ...         ...        ...
588                        Кувандык           0          0
589                        Куйбышев           0          0
590  Куйбышево (Ростовская область)           0          0
591                        Култаево           0          0
592                        Кульсары           0          0

[593 rows x 3 columns]

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [27]:
query_5_3 = f'''select
                e.id,
                count(distinct v.area_id) areas
            from public.employers e
                LEFT JOIN public.vacancies v on v.employer_id = e.id
            GROUP BY e.id
            ORDER BY count(v.area_id) desc
            '''

In [28]:
df_5_3 = pd.read_sql_query(query_5_3, connection)
display(df_5_3)

C:\Users\kab27\AppData\Local\Temp\ipykernel_12832\596095100.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_5_3 = pd.read_sql_query(query_5_3, connection)


id  areas
0         1740    181
1         2748    152
2        78638     43
3         3529     24
4        39305     24
...        ...    ...
23496  5970935      0
23497  5524047      0
23498     3601      0
23499  5971105      0
23500      320      0

[23501 rows x 2 columns]

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [29]:
query_5_4 = f'''select
                count(distinct e.id)
            from public.employers e
                LEFT JOIN public.employers_industries ei on ei.employer_id = e.id
            WHERE industry_id is null
            '''

In [30]:
df_5_4 = pd.read_sql_query(query_5_4, connection)
display(df_5_4)

C:\Users\kab27\AppData\Local\Temp\ipykernel_12832\3722588189.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_5_4 = pd.read_sql_query(query_5_4, connection)


count
0   8419

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [31]:
query_5_5 = f'''select
                e.name
            from public.employers e
                LEFT JOIN public.employers_industries ei on ei.employer_id = e.id
            GROUP BY e.name
            HAVING COUNT(ei.industry_id)=4
            Order BY 1
            OFFSET 2 LIMIT 1
            '''

In [32]:
df_5_5 = pd.read_sql_query(query_5_5, connection)
display(df_5_5)

C:\Users\kab27\AppData\Local\Temp\ipykernel_12832\3881363819.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_5_5 = pd.read_sql_query(query_5_5, connection)


name
0  2ГИС

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [33]:
query_5_6 = f'''select
                COUNT(distinct e.id)
            from public.employers e
                LEFT JOIN public.employers_industries ei on ei.employer_id = e.id
                LEFT JOIN public.industries i on i.id = ei.industry_id
            WHERE i.name='Разработка программного обеспечения'
            '''

In [34]:
df_5_6 = pd.read_sql_query(query_5_6, connection)
display(df_5_6)

C:\Users\kab27\AppData\Local\Temp\ipykernel_12832\2900473628.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_5_6 = pd.read_sql_query(query_5_6, connection)


count
0   3553

7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [35]:
url = 'https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8'
table = pd.read_html(url)
mill_cities = pd.DataFrame(table[1])
mill_cities = tuple(mill_cities['Город'].replace('Воронеж[a]', 'Воронеж'))
mill_cities






('Москва',
 'Санкт-Петербург',
 'Новосибирск',
 'Екатеринбург',
 'Казань',
 'Нижний Новгород',
 'Красноярск',
 'Челябинск',
 'Самара',
 'Уфа',
 'Ростов-на-Дону',
 'Омск',
 'Краснодар',
 'Воронеж',
 'Пермь',
 'Волгоград')

In [36]:
query_5_7 = f'''select
                    a.name,
                    count(*) cnt
                from public.employers e
                JOIN public.vacancies v on v.employer_id = e.id
                JOIN public.areas a on a.id = v.area_id
                WHERE a.name in {mill_cities} and e.name = 'Яндекс'
                GROUP BY a.id
                
                UNION ALL
                
                select
                    'Total', count(*)
                from public.vacancies v
                JOIN public.employers e ON e.id = v.employer_id
                JOIN public.areas a ON a.id = v.area_id
                WHERE a.name in {mill_cities} and e.name = 'Яндекс'
                '''

In [37]:
df_5_7 = pd.read_sql_query(query_5_7, connection)
display(df_5_7)

C:\Users\kab27\AppData\Local\Temp\ipykernel_12832\3666685428.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_5_7 = pd.read_sql_query(query_5_7, connection)


name  cnt
0             Total  485
1            Москва   54
2   Санкт-Петербург   42
3      Екатеринбург   39
4       Новосибирск   35
5         Волгоград   24
6           Воронеж   32
7         Краснодар   30
8        Красноярск   23
9   Нижний Новгород   36
10             Омск   21
11            Пермь   25
12   Ростов-на-Дону   25
13           Самара   26
14           Казань   25
15              Уфа   26
16        Челябинск   22

***

Выводы:
1. Наибольшее количество вакансий размещено в крупных городах
2. У некоторых работадателей нет опубликованных вакансий
3. Некоторый работадатели публикуют вакансии во многих регионах при этом есть работадатели которые не указывают ни одного региона.
4. Много работадателей которые не указывают сферы деятельности, есть работадатели которые указывают несколько сфер деятельности.
5. IT компании публикуют много вакансий в большом количестве регионов

# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [38]:
query_6_1 = f'''select
                name
            from public.vacancies
            WHERE name ilike '%data%' 
            or name ilike '%данн%'
            '''

In [39]:
df_6_1 = pd.read_sql_query(query_6_1, connection)
display(df_6_1)

C:\Users\kab27\AppData\Local\Temp\ipykernel_12832\43016786.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_6_1 = pd.read_sql_query(query_6_1, connection)


name
0     Аналитик хранилищ данных и BI-систем/Аналитик ...
1                              Data-analysts (удаленно)
2     Стажер Data Engineer/Data Аналитик в команду Б...
3                    Старший специалист хранилищ данных
4                      Администратор баз данных и сетей
...                                                 ...
1766        Программист Delphi / разработчик баз данных
1767                  Разработчик баз данных (Big Data)
1768                                    Аналитик данных
1769              Data аналитик /Аналитик BI (удаленно)
1770                                   DataOps Engineer

[1771 rows x 1 columns]

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [40]:
query_6_2 = f'''select
                name
            from public.vacancies
            WHERE 
            (name ilike '%data scientist%' 
            or name ilike '%data science%'
            or name ilike '%исследователь данных%'
            or (name like '%ML%' and name not ilike '%HTML%')
            or name ilike '%machine learning%'
            or name ilike '%машинн%обучен%')
            and (name ilike '%junior%' 
            or experience = 'Нет опыта' 
            or employment = 'Стажировка')
            '''

In [41]:
df_6_2 = pd.read_sql_query(query_6_2, connection)
display(df_6_2)

C:\Users\kab27\AppData\Local\Temp\ipykernel_12832\517244489.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_6_2 = pd.read_sql_query(query_6_2, connection)


name
0                             Junior+ CV/ML developer
1                            Data Scientist (Junior+)
2                      Junior Data Scientist (Стажер)
3      Математик, Junior/Middle/Senior Data Scientist
4                               Junior Data Scientist
5   Data Scientist/Machine Learning Engineer (Deve...
6                             Data Scientist (Junior)
7                        Intern/Junior Data Scientist
8                          Junior Data Scientist (VK)
9                        Junior/Middle Data Scientist
10          Data Scientist (Computer Vision) - Junior
11  Исследователь в области машинного обучения и м...
12                             Data Scientist, Junior
13  Junior Data Scientist Управления модельных рис...
14  Математик-аналитик (data scientist / data anal...
15                    Data Scientist (junior+/middle)
16                              Junior Data Scientist
17                 Research Engineer / Data Scientist
18  Стажер направления Машинное обучение и анализ ...
19  Стажёр Data Scientist / Группа сервисов Антифр...
20                              Junior Data Scientist
21          Senior Research Engineer / Data Scientist
22                              Ведущий AI/ML инженер
23        Data scientist (Junior/Middle) в консалтинг
24  Junior Data Scientist (математик-аналитик) уда...
25             Junior Data Scientist (Promo Forecast)
26  Data Scientist/Machine Learning Engineer (Deve...
27                              Data Scientist в Сбер
28  Data Scientist/Machine Learning Engineer (Deve...
29  Junior Data Scientist (Валидация скоринговых м...
30  Junior Data Scientist (Валидация скоринговых м...
31  Junior Data Scientist (Валидация скоринговых м...
32  Junior Data Scientist (Валидация скоринговых м...
33  Junior Data Scientist (Валидация скоринговых м...
34  Junior Data Scientist (Валидация скоринговых м...
35  Junior Data Scientist (Валидация скоринговых м...
36  Junior Data Scientist (Валидация скоринговых м...
37  Junior Data Scientist (Валидация скоринговых м...
38            Риск - Аналитик (Junior Data scientist)
39  Программист Python/ Data Scientist (junior / m...
40       Риск-аналитик (ML Engineer / Data Scientist)
41            Data Scientist/ML engineer (Python/C++)
42       Junior Data scientist/ Младший риск-аналитик
43                         Senior Data Scientist (AI)
44             Junior Data Scientist for NLP (Python)
45          Junior Research Engineer / Data Scientist
46                              Ведущий AI/ML инженер
47                                     Data Scientist
48              Python developer / Junior ML engineer
49                             Data Scientist, Junior
50          Data Scientist (Computer Vision) / Junior

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [42]:
query_6_3 = f'''select
                name
            from public.vacancies
            WHERE 
            (name ilike '%data%scientist%' 
            or name ilike '%data%science%'
            or name ilike '%исследователь%данных%'
            or (name like '%ML%' and name not ilike '%HTML%')
            or name ilike '%machine%learning%'
            or name ilike '%машинн%обучен%%')
            and (key_skills ilike '%SQL%' or key_skills ilike '%postgres%')
            '''

In [43]:
df_6_3 = pd.read_sql_query(query_6_3, connection)
display(df_6_3)

C:\Users\kab27\AppData\Local\Temp\ipykernel_12832\352076616.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_6_3 = pd.read_sql_query(query_6_3, connection)


name
0                           Junior+ CV/ML developer
1                                    Data Scientist
2                                    Data Scientist
3                                    Data Scientist
4                                    Data Scientist
..                                              ...
196                          Data Scientist, Junior
197       Data Scientist (Computer Vision) / Junior
198       Data Scientist (Computer Vision) / Middle
199  Machine Learning and Computer vision developer
200                   Middle+/Senior Data Scientist

[201 rows x 1 columns]

4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [44]:
query_6_4 = f'''select
                name
            from public.vacancies
            WHERE 
            (name ilike '%data%scientist%' 
            or name ilike '%data%science%'
            or name ilike '%исследователь%данных%'
            or (name like '%ML%' and name not ilike '%HTML%')
            or name ilike '%machine%learning%'
            or name ilike '%машинн%обучен%%')
            and (key_skills ilike '%python%')
            '''

In [45]:
df_6_4 = pd.read_sql_query(query_6_4, connection)
display(df_6_4)

C:\Users\kab27\AppData\Local\Temp\ipykernel_12832\406769518.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_6_4 = pd.read_sql_query(query_6_4, connection)


name
0                              Junior+ CV/ML developer
1                                       Data Scientist
2                             Data Scientist (Junior+)
3                                AI/ML Senior Engineer
4                                       Data Scientist
..                                                 ...
346     Machine Learning and Computer vision developer
347  Data scientist / Исследователь-программист данных
348                                   Специалист ML/DS
349   Data Scientist в области рекомендательных систем
350                  ML Engineer Team Lead (to Cyprus)

[351 rows x 1 columns]

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [46]:
query_6_5 = f'''select
                ROUND(AVG(1 + length(key_skills) - length(Replace(key_skills, CHR(9), ''))), 2)
            from public.vacancies
            WHERE 
            (name ilike '%data%scientist%' 
            or name ilike '%data%science%'
            or name ilike '%исследователь%данных%'
            or (name like '%ML%' and name not ilike '%HTML%')
            or name ilike '%machine%learning%'
            or name ilike '%машинн%обучен%%')
            '''

In [47]:
df_6_5 = pd.read_sql_query(query_6_5, connection)
display(df_6_5)

C:\Users\kab27\AppData\Local\Temp\ipykernel_12832\1427856002.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_6_5 = pd.read_sql_query(query_6_5, connection)


round
0   6.41

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [48]:
query_6_6 = f'''select
                distinct(experience),
                ROUND(AVG((coalesce(salary_to, salary_from) + coalesce(salary_from, salary_to))/2))
            from public.vacancies
            WHERE 
            (name ilike '%data%scientist%' 
            or name ilike '%data%science%'
            or name ilike '%исследователь%данных%'
            or (name like '%ML%' and name not ilike '%HTML%')
            or name ilike '%machine%learning%'
            or name ilike '%машинн%обучен%%')
            and (salary_to is not null or salary_from is not null)
            and experience = 'От 3 до 6 лет'
            GROUP BY experience
            '''

In [49]:
df_6_6 = pd.read_sql_query(query_6_6, connection)
display(df_6_6)

C:\Users\kab27\AppData\Local\Temp\ipykernel_12832\3008645249.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_6_6 = pd.read_sql_query(query_6_6, connection)


experience     round
0  От 3 до 6 лет  243115.0

In [81]:
query_0_1 = f'''select
                e.name as "Company name",
                count(v.id) as "Vacancies count"
            from public.vacancies v
            LEFT JOIN employers e ON e.id = v.employer_id
            WHERE 
            (v.name ilike '%data%scientist%' 
            or v.name ilike '%data%science%'
            or v.name ilike '%исследователь%данных%'
            or (v.name like '%ML%' and v.name not ilike '%HTML%')
            or v.name ilike '%machine%learning%'
            or v.name ilike '%машинн%обучен%%')
            GROUP BY e.name
            ORDER BY count(v.id) desc
            LIMIT 10
            '''

In [82]:
df_0_1 = pd.read_sql_query(query_0_1, connection)
display(df_0_1)

C:\Users\kab27\AppData\Local\Temp\ipykernel_12832\3667547203.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_0_1 = pd.read_sql_query(query_0_1, connection)


Company name  Vacancies count
0                   СБЕР               38
1        Bell Integrator               25
2         Банк ВТБ (ПАО)               18
3                     VK               15
4  Positive Technologies               11
5                 Яндекс                9
6         EvenBet Gaming                9
7                МегаФон                8
8               Andersen                7
9                   Ozon                6

In [83]:
connection.close()

***

На площадке HeadHunter размещено много вакансий связанных по работе с данными. Вакансий относящихся к DS почти в 35 раз меньше. Для работы по специальности DS нужно владеть в среднем 6-7 ключевыми навыками (одним из которых являеться SQL). При этом SQL является ключевым навыком в 201 вакансии. В зависимости от опыта работ специалист по DS может расчитывать на среднюю за от 74 000 до 240 000.

# Общий вывод по проекту

В целом данные данные соответствуют целям проекта и позволяют провести анализ и выбрать вакансии удовлетворяющие условиям проекта.

Среди представленной выборки вакансий вакансии связанные с работой с данными занимают 3,7% а вакансиий по DS еще меньше. Не смотря на это можно провести выборку в зависимости от региона, опыта работы сооискателя и желаемых графика работы и типа трудоустройства.

